### PR TITLE
Fix JSON cast in TRY expression

### PR DESCRIPTION
**Pull request by kagamiori**
Summary:
When evaluating a `TRY(CAST(... as JSON))` expression from MAP or ARRAY
inputs, exceptions from the CAST operation are kept in an error vector in EvalCtx
that will be used to null out the result of the TRY expression. Before this change, the
error vector is automatically populated in EvalCtx::applyToSelectedNoThrow() loops
over the elements of array or map, such that the rows in the error vector doesn't map
to the top-level rows of the array or map vector.

This diff fixes this issue by keeping a reverse mapping from element rows to top-level
rows of the map or array, and manually insert exceptions into the error vector at the
correct position corresponding to the reverse mapping.

This diff also replace EvalCtx::applyToSelectedNoThrow() loops in JsonType.cpp that
do not perform cast operation on input data with SelectivityVector::applyToSelected()
instead, because those are Velox internal logic and should throw when there are
unexpected errors.

Differential Revision: D38546082

### CODE REVIEW

File: velox/functions/prestosql/types/JsonType.cpp [hunk 1]
```diff
@@ -245,15 +312,20 @@ void castToJsonFromArray(
...
+
+  AsJson elementsAsJson{
+      context, elements, elementsRows, elementsToTopLevelRows};
 
   // Estimates an upperbound of the total length of all Json strings for the
   // input according to the length of all elements Json strings and the
   // delimiters to be added.
   size_t elementsStringSize = 0;
-  context->applyToSelectedNoThrow(rows, [&](auto row) {
+  rows.applyToSelected([&](auto row) {
```
Comment (pedroerp) [inline velox/functions/prestosql/types/JsonType.cpp:328]: I might be missing something, but looks like this pattern is common enough that will (or probably already is) cause bugs in other parts of the codebase that rely on `applyToSelectedNoThrow()` over non-primitive types, right?

I wonder if we should come up with something more generic, maybe like a `applyToSelectedArrayNoThrow()` helper functions that takes a lambda and passes both `row` and `elemRow` to it automatically? Thoughts?

Comment (kagamiori) [reply velox/functions/prestosql/types/JsonType.cpp:328]: You're absolutely right! The same issue exists in CastExpr::applyArray() and CastExpr::applyMap (e.g., cast(array(varchar) as array(bigint))) that I plan to fix later. The idea of `applyToSelectedArrayNoThrow()` sounds good. I'll think about it. Thank you for this suggestion!

Comment (pedroerp) [reply velox/functions/prestosql/types/JsonType.cpp:328]: Sounds good. I'm thinking that applyToSelectedNoThrow() is also used in many vectorized functions outside of Casts, so this bug might also be affecting some of them.

Comment (kagamiori) [reply velox/functions/prestosql/types/JsonType.cpp:328]: Ouch...

Comment (kagamiori) [reply velox/functions/prestosql/types/JsonType.cpp:328]: @pedroerp So I've thought about making a `applyToSelectedArrayNoThrow()` for this fix, but there is a difficulty for casting arrays or maps. The problem is that when looping over the element vector(s), we not only translate the SelectivityVector of top-level rows to element rows, but also translate rows of the decoded vector to rows of its base vector. So we need to reverse both transformation to map the row where an exception throws back to the corresponding top-level row. The base-row to decoded-row mapping is a one-to-many that cannot be captured by a simple data structure like Buffer. (The current code instead loops over the decoded-rows and check whether there are exceptions at the corresponding base-rows.) Please let me know if you have any suggestions.

Comment (kagamiori) [reply velox/functions/prestosql/types/JsonType.cpp:328]: Updated by moving the common logic to EvalCtx::applyToSelectedElementsNoThrow() and EvalCtx::applyToSelectedPeeledElementsNoThrow(). @pedroerp

**Comment by kagamiori** [comment]
Closing this PR because it has been fixed in https://github.com/facebookincubator/velox/pull/2760 and https://github.com/facebookincubator/velox/pull/2761.

**Pull request closed without merging**